### PR TITLE
Enable type reduction for Gather CPU kernel.

### PR DESCRIPTION
**Pull request by edgchen1**
**Description**
Enable type reduction for Gather CPU kernel.

**Motivation and Context**
Allows implementation to be excluded from binary if the type is not enabled.

### CODE REVIEW

File: onnxruntime/core/providers/op_kernel_type_control_utils.h
Comment (edgchen1) [inline onnxruntime/core/providers/op_kernel_type_control_utils.h:1]: copied from https://github.com/microsoft/onnxruntime/pull/6547
will merge after that PR is merged

**Comment by skottmckay** [comment]
> }

nit: message should probably say 'not supported in this build' as we support both index types that are valid according to the Gather spec.

---
Refers to: onnxruntime/core/providers/cpu/tensor/gather.cc:159 in 63f1fb2. [](commit_id = 63f1fb2702ea5b6f00e42b2c8fa84c7479f89088, deletion_comment = False)

**Review by skottmckay — DISMISSED** [review]
:shipit:

**Review by skottmckay — APPROVED** [review]

**Pull request merged**
